README.md

```diff
@@ -632,9 +632,17 @@ arguments: the session data, and a callback function.
 
 ```javascript
 
-const bearerAuthorizer = (data, block) => {
+const bearerAuthorizer = (data, headers, options, url) => {
   if (data.token) {
-    block('Authorization', `Bearer ${data.token}`)
+    headers('Authorization', `Bearer ${data.token}`)
+  }
+  // Or perhaps ...
+  if (data.usingCookieAuth) {
+    options('credentials', `same-origin`)
+  }
+  // Or perhaps ...
+  if (data.url !== '/my-unprotected-api') {
```

Comment:
Seeing this API in the example gives me some hesitation shipping it this way. I'm afraid too many users may forget which order these arguments appear. On top of that, it makes it hard to extend this API in some future event we decide we need to expose more options to the authorizer function. 

For historical context, this API was originally derived from [Ember Simple Auth's implementation](http://ember-simple-auth.com/api/classes/BaseAuthorizer.html). Note the 2 arguments.

That being said, and the more I think about it, I think we are headed in the wrong direction by adding this logic to the authorizer. I still think it makes the most sense that we manipulate headers-only in the authorizer.

I think instead, we should add an additional option to the middleware itself. Maybe something like `transformRequest`? The idea here is that it would be a final stop to allow the user to modify any last minute options before the request is sent. I think this makes a bit more logical sense for things like manipulating the `credentials` option and for other options that may or may not be related to authorization purposes.

Here is a rough idea (by all means, I'd love for you to help drive this API)

```js
const authMiddleware = createAuthMiddleware({
  authorize: (data, header) => { /* ... */ },
  transformRequest: (options) => ({
    ...options, // options.headers would include any headers added via `authorize`
    credentials: 'same-origin'
  })
})
```

In this case, the `transformRequest` function would need to return an object containing the request options that will be used in the final `fetch` call. I think the thing here is figuring out exactly what to pass `transformRequest`. I omitted `data` in this example, but this may very much be a driving factor in the need for this option, so maybe we also pass that?

I'd love to get your thoughts on how to drive this API and how best to serve the use cases.

Reply:
Yea, I think the `transformRequest` middleware option is a cleaner approach. The P.R I opened has conflated two distinct issues so its possibly best if we address those requirements seperate.

Requirement 1: **More control over the fetch options**. `transformRequest` would meet that requirement.

Requirement 2. @ogenodisho (a colleague of mine) requirement is probably more related to the conversation we had here: https://github.com/jerelmiller/redux-simple-auth/pull/31#issuecomment-402907113. E.g. the internal app framework we use in our company is based around a component based architecture and the component does not know about authentication or how that happened, it just has a hook to use a `fetch`.

So by passing the URL to the authenticator, this allows the parent app host to inspect the request url and decide *IF* it should add auth headers or not. e.g. it removes the need for a component to know what URLs require authentication. I think this requirement is driven more from our custom application architecture, rather then a distinct limitation with the redux-simple-auth middleware.

I wonder if this requirement would be better handled by having multiple auth contexts supported in this library. So its almost like an application can map a URL domain to an authenticator/authoriser (during authentication). this would mean that redux-simple-auth can manage multiple different auth tokens inside an app.

From a quick glance at the code (https://github.com/jerelmiller/redux-simple-auth/blob/master/packages/redux-simple-auth/src/middleware.js#L61) - seems that its possible to have multiple authenticators configured, BUT its only possible to have a single authoriser. If the middleware allowed supporting 1 authenticator to 1 authoriser - then I think this would allow requirement 2. to be met. 

e.g. the app can just register the correct authenticator/authoriser for the different APIs it needs to call.

Reply:
Multiple auth contexts are definitely something that I would love to explore. That makes a lot of sense to me.

I'd say lets move forward with `transformRequest` as I think it would be the best bet to solve these issues in the interim. I think there is a bit more thinking in the API design of multiple contexts that I'd like to do before moving forward on that solution.

Thanks for the wonderful feedback!